File: paykit-mobile/ios-demo/PaykitDemo/PaykitDemo/PaykitDemo/Storage/ContactStorage.swift
//
//  ContactStorage.swift
//  PaykitDemo
//
//  Persistent storage for contacts using Keychain.
//

import Foundation

/// Manages persistent storage of contacts
class ContactStorage {
    
    private let keychain: KeychainStorage
    private let identityName: String
    
    // In-memory cache
    private var contactsCache: [Contact]?
    
    private var contactsKey: String {
        "paykit.contacts.\(identityName)"
    }
    
    init(identityName: String, keychain: KeychainStorage = KeychainStorage(serviceIdentifier: "com.paykit.demo")) {
        self.identityName = identityName
        self.keychain = keychain
    }
    
    // MARK: - CRUD Operations
    
    /// Get all contacts
    func listContacts() -> [Contact] {
        if let cached = contactsCache {
            return cached
        }
        
        do {
            guard let data = try keychain.retrieve(key: contactsKey) else {
                return []
            }
            let contacts = try JSONDecoder().decode([Contact].self, from: data)
            contactsCache = contacts
            return contacts
        } catch {
            print("ContactStorage: Failed to load contacts: \(error)")
            return []
        }
    }
    
    /// Get a specific contact
    func getContact(id: String) -> Contact? {
        return listContacts().first { $0.id == id }
    }
    
    /// Save a new contact or update existing
    func saveContact(_ contact: Contact) throws {
        var contacts = listContacts()
        
        if let index = contacts.firstIndex(where: { $0.id == contact.id }) {
            // Update existing
            contacts[index] = contact
        } else {
            // Add new
            contacts.append(contact)
        }
        
        try persistContacts(contacts)
    }
    
    /// Delete a contact
    func deleteContact(id: String) throws {
        var contacts = listContacts()
        contacts.removeAll { $0.id == id }
        try persistContacts(contacts)
    }
    
    /// Search contacts by name
    func searchContacts(query: String) -> [Contact] {
        let query = query.lowercased()
        return listContacts().filter { contact in
            contact.name.lowercased().contains(query) ||
            contact.publicKeyZ32.lowercased().contains(query)
        }
    }
    
    /// Record a payment to a contact
    func recordPayment(contactId: String) throws {
        var contacts = listContacts()
        guard let index = contacts.firstIndex(where: { $0.id == contactId }) else {
            return
        }
        
        contacts[index].recordPayment()
        try persistContacts(contacts)
    }
    
    /// Clear all contacts
    func clearAll() throws {
        try persistContacts([])
    }
    
    /// Import contacts (merge with existing)
    func importContacts(_ newContacts: [Contact]) throws {
        var contacts = listContacts()
        
        for newContact in newContacts {
            if !contacts.contains(where: { $0.id == newContact.id }) {
                contacts.append(newContact)
            }
        }
        
        try persistContacts(contacts)
    }
    
    /// Export contacts as JSON string
    func exportContacts() throws -> String {
        let contacts = listContacts()
        let data = try JSONEncoder().encode(contacts)
        return String(data: data, encoding: .utf8) ?? "[]"
    }
    
    // MARK: - Private
    
    private func persistContacts(_ contacts: [Contact]) throws {
        let data = try JSONEncoder().encode(contacts)
        try keychain.store(key: contactsKey, data: data)
        contactsCache = contacts
    }
}

// MARK: - Errors

enum ContactStorageError: LocalizedError {
    case encodingFailed
    case decodingFailed
    case notFound(id: String)
    
    var errorDescription: String? {
        switch self {
        case .encodingFailed:
            return "Failed to encode contacts"
        case .decodingFailed:
            return "Failed to decode contacts"
        case .notFound(let id):
            return "Contact not found: \(id)"
        }
    }
}

File: paykit-mobile/android-demo/app/src/main/java/com/paykit/demo/storage/ContactStorage.kt
package com.paykit.demo.storage

import android.content.Context
import android.util.Log
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import com.paykit.demo.model.Contact
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

/**
 * Manages persistent storage of contacts using EncryptedSharedPreferences.
 */
class ContactStorage(context: Context) {
    
    companion object {
        private const val TAG = "ContactStorage"
        private const val PREFS_NAME = "paykit_contacts"
        private const val CONTACTS_KEY = "contacts_list"
    }
    
    private val json = Json { 
        ignoreUnknownKeys = true 
        encodeDefaults = true
    }
    
    private val prefs by lazy {
        try {
            val masterKey = MasterKey.Builder(context)
                .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
                .build()
            
            EncryptedSharedPreferences.create(
                context,
                PREFS_NAME,
                masterKey,
                EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
                EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
            )
        } catch (e: Exception) {
            Log.e(TAG, "Failed to create encrypted prefs, falling back to regular prefs", e)
            context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        }
    }
    
    // In-memory cache
    private var contactsCache: List<Contact>? = null
    
    // MARK: - CRUD Operations
    
    /**
     * Get all contacts
     */
    fun listContacts(): List<Contact> {
        contactsCache?.let { return it }
        
        return try {
            val jsonString = prefs.getString(CONTACTS_KEY, null) ?: return emptyList()
            val contacts = json.decodeFromString<List<Contact>>(jsonString)
            contactsCache = contacts
            contacts
        } catch (e: Exception) {
            Log.e(TAG, "Failed to load contacts: ${e.message}")
            emptyList()
        }
    }
    
    /**
     * Get a specific contact
     */
    fun getContact(id: String): Contact? {
        return listContacts().find { it.id == id }
    }
    
    /**
     * Save a new contact or update existing
     */
    fun saveContact(contact: Contact) {
        val contacts = listContacts().toMutableList()
        val index = contacts.indexOfFirst { it.id == contact.id }
        
        if (index >= 0) {
            contacts[index] = contact
        } else {
            contacts.add(contact)
        }
        
        persistContacts(contacts)
    }
    
    /**
     * Delete a contact
     */
    fun deleteContact(id: String) {
        val contacts = listContacts().toMutableList()
        contacts.removeAll { it.id == id }
        persistContacts(contacts)
    }
    
    /**
     * Search contacts by name or public key
     */
    fun searchContacts(query: String): List<Contact> {
        val lowerQuery = query.lowercase()
        return listContacts().filter { contact ->
            contact.name.lowercase().contains(lowerQuery) ||
            contact.publicKeyZ32.lowercase().contains(lowerQuery)
        }
    }
    
    /**
     * Record a payment to a contact
     */
    fun recordPayment(contactId: String) {
        val contacts = listContacts().toMutableList()
        val index = contacts.indexOfFirst { it.id == contactId }
        
        if (index >= 0) {
            contacts[index] = contacts[index].recordPayment()
            persistContacts(contacts)
        }
    }
    
    /**
     * Clear all contacts
     */
    fun clearAll() {
        persistContacts(emptyList())
    }
    
    /**
     * Import contacts (merge with existing)
     */
    fun importContacts(newContacts: List<Contact>) {
        val contacts = listContacts().toMutableList()
        
        for (newContact in newContacts) {
            if (contacts.none { it.id == newContact.id }) {
                contacts.add(newContact)
            }
        }
        
        persistContacts(contacts)
    }
    
    /**
     * Export contacts as JSON string
     */
    fun exportContacts(): String {
        return try {
            json.encodeToString(listContacts())
        } catch (e: Exception) {
            Log.e(TAG, "Failed to export contacts: ${e.message}")
            "[]"
        }
    }
    
    // MARK: - Private
    
    private fun persistContacts(contacts: List<Contact>) {
        try {
            val jsonString = json.encodeToString(contacts)
            prefs.edit().putString(CONTACTS_KEY, jsonString).apply()
            contactsCache = contacts
        } catch (e: Exception) {
            Log.e(TAG, "Failed to save contacts: ${e.message}")
        }
    }
}

